File: haskell/src/Day02.hs
module Day02 (day02) where

import qualified Text.Parsec as Parsec

data Entry = Entry Int Int Char String
instance Show Entry where
    show (Entry n1 n2 ch pw) = "Entry: " ++ show n1 ++ ", " ++ show n2 ++ " (" ++ [ch] ++ "): " ++ pw

day02 :: IO ()
day02 = do
    s <- readFile "../data/day02.txt"
    let entries = parseData s
    let answer1 = length $ filter valid1 entries
    let answer2 = length $ filter valid2 entries
    putStrLn "\n===== Day 2 ====="
    putStrLn $ "The number of valid passwords with the old policy is " ++ show answer1
    putStrLn $ "The number of valid passwords with the new policy is " ++ show answer2

valid1 :: Entry -> Bool
valid1 (Entry n1 n2 ch pw) =
    matchingChars >= n1 && matchingChars <= n2
    where
        matchingChars = length $ filter (ch ==) pw 

valid2 :: Entry -> Bool
valid2 (Entry n1 n2 ch pw) =
    (ch == pw!!(n1-1)) /= (ch == pw!!(n2-1))

parseData :: String -> [Entry]
parseData =
    map parseLine . lines

parseLine :: String -> Entry
parseLine s =
    case Parsec.parse entryParse "" s of
        Right entry -> entry
        Left _ -> Entry 0 0 ' ' "Nothing"

entryParse :: Parsec.Parsec String () Entry
entryParse = do
    n1 <- parseInt
    Parsec.char '-'
    n2 <- parseInt
    Parsec.spaces
    ch <- Parsec.anyChar
    Parsec.string ": "
    pw <- Parsec.many1 Parsec.anyChar
    return $ Entry n1 n2 ch pw

parseInt :: Parsec.Parsec String () Int
parseInt = do
    s <- Parsec.many1 Parsec.digit
    return (read s :: Int)

File: typescript/src/day02.ts
import { count } from './util/array';

interface Entry {
  n1: number;
  n2: number;
  char: string;
  password: string;
}

export default function day2(data: string) {
  const entries = parseData(data);
  const answer1 = count(entries, valid1);
  const answer2 = count(entries, valid2);
  console.log('====== Day 2 ======')
  console.log('The number of valid passwords with the old policy is', answer1);
  console.log('The number of valid passwords with the new policy is', answer2);
}


export function parseData(s: string): Entry[] {
  const lines = s.split('\n');
  return lines.map(line => {
    const [_, n1, n2, char, password] = line.match(/(\d+)-(\d+) (.): (.+)/) || [];
    return { n1: +n1, n2: +n2, char, password }
  })
}

function valid1(entry: Entry): boolean {
  const nrOfChars = count(entry.password.split(''), (ch) => ch === entry.char);
  return nrOfChars >= entry.n1 && nrOfChars <= entry.n2;
}

function valid2(entry: Entry): boolean {
  const xor = (b1, b2) => b1 ? !b2 : b2;
  const { n1, n2, char, password } = entry;
  return xor(password[n1-1] === char, password[n2-1] === char);
}